source/websocket_decoder.c

```diff
@@ -0,0 +1,331 @@
+
+        if (decoder->expecting_continuation_data_frames != is_continuation_frame) {
+            return aws_raise_error(AWS_ERROR_HTTP_PARSE);
+        }
+
+        decoder->expecting_continuation_data_frames = !decoder->current_frame.fin;
+
+    } else {
+        /* Control frames themselves MUST NOT be fragmented. */
+        if (!decoder->current_frame.fin) {
+            return aws_raise_error(AWS_ERROR_HTTP_PARSE);
+        }
+    }
+
+    decoder->state++;
+    return AWS_OP_SUCCESS;
+}
+
+/* STATE_LENGTH_BEGIN: Decode byte containing length, determine if we need to decode extended length. */
+static int s_state_length_begin(struct aws_websocket_decoder *decoder, struct aws_byte_cursor *data) {
+    if (data->len == 0) {
+        return AWS_OP_SUCCESS;
+    }
+
+    uint8_t byte = data->ptr[0];
+    aws_byte_cursor_advance(data, 1);
+
+    /* first bit is a bool */
+    decoder->current_frame.masked = byte & 0x80;
+
+    /* remaining 7 bits are payload length */
+    decoder->current_frame.payload_length = byte & 0x7F;
+
+    /* If 7bit payload length has a high value, then the next few bytes contain the real payload length */
+    if (decoder->current_frame.payload_length < VALUE_FOR_2BYTE_EXTENDED_LENGTH) {
+        decoder->state += 2; /* Skip next state, which would have processed extended length */
+    } else {
+        decoder->state_bytes_processed = 0;
+        decoder->state++;
+    }
+
+    return AWS_OP_SUCCESS;
+}
+
+/* STATE_LENGTH_CONTINUE: Decode extended length (skipped if no extended length) . */
+static int s_state_length_continue(struct aws_websocket_decoder *decoder, struct aws_byte_cursor *data) {
+    if (data->len == 0) {
+        return AWS_OP_SUCCESS;
+    }
+
+    /* RFC-6455 Section 5.2 Base Framing Protocol - Payload length
+     * 1) It is a crime to use more bytes than necessary to encode length.
+     *      > in all cases, the minimal number of bytes MUST be used to encode
+     *      > the length, for example, the length of a 124-byte-long string
+     *      > can't be encoded as the sequence 126, 0, 124
+     * 2) It is a crime to use most-significant bit on 8 byte payloads.
+     *      > If 127, the following 8 bytes interpreted as a 64-bit unsigned integer
+     *      > (the most significant bit MUST be 0) are the payload length */
+    uint8_t total_bytes_extended_length;
+    uint64_t min_acceptable_value;
+    uint64_t max_acceptable_value;
+    if (decoder->current_frame.payload_length == VALUE_FOR_2BYTE_EXTENDED_LENGTH) {
+        total_bytes_extended_length = 2;
+        min_acceptable_value = VALUE_FOR_2BYTE_EXTENDED_LENGTH;
+        max_acceptable_value = UINT16_MAX;
+    } else {
+        assert(decoder->current_frame.payload_length == VALUE_FOR_8BYTE_EXTENDED_LENGTH);
+
+        total_bytes_extended_length = 8;
+        min_acceptable_value = UINT16_MAX + 1;
+        max_acceptable_value = 0x8000000000000000ULL - 1;
```

Comment:
Why not just INT64_MAX?

Reply:
Spec says it's a crime to set the high bit, even though it should be treated as an unsigned u64